### PR TITLE
feat: support configure_fit() in TFKerasTrial [DET-4528]

### PR DESCRIPTION
**Pull request by rb-determined-ai**
## Description

Users can call context.configure_fit() to set additional arguments in
the TFKerasTrial API that were previously only configurable via the
Native API.

## Test Plan

Modified a unit test.

## Commentary

Overall, this PR has been pulled out of the keras data cleanup work (#1545 or #1564) to ensure that it lands within the same release as the improved callback support (so that the `self.context.configure_fit(verbose=true)` is the preferred API from the start).  That's relevant because this PR actually makes a lot more sense on top of the other changes, even though it is functionally correct here as well.

`shuffle` will be supported after fixing up the keras data handling.

### CODE REVIEW

File: harness/determined/keras/_tf_keras_context.py [hunk 1]
```diff
@@ -40,6 +49,61 @@ def __init__(self, env: det.EnvContext, hvd_config: horovod.HorovodContext):
...
+
+        **Usage Example**
+
+        .. code::
+
+           class MyTFKerasTrial(det.keras.TFKerasTrial):
+               def __init__(self, context):
+                   ...
+                   self.context.configure_fit(verbose=False, workers=5)
+                   self.context.configure_fit(use_multiprocessing=True)
```
Comment (aaron276h) [inline harness/determined/keras/_tf_keras_context.py:94]: non-blocking: why not include all these args in one call to `configure_fit()`?

File: harness/determined/keras/_tf_keras_context.py [hunk 2]
```diff
@@ -40,6 +49,61 @@ def __init__(self, env: det.EnvContext, hvd_config: horovod.HorovodContext):
...
+
+           -  ``class_weight``: defaults to ``None``
+
+           -  ``workers``: defaults to 1. Ignored for ``tf.data.dataset`` data.
+
+           -  ``use_multiprocessing``: defaults to `False`. Ignored for ``tf.data.dataset`` data.
+
+           -  ``max_queue_size``: defualts to `10`. Ignored for ``tf.data.dataset`` data.
+
+        Note that any keyword arguments which are not provided will not overwrite any previously
```
Comment (aaron276h) [inline harness/determined/keras/_tf_keras_context.py:83]: non-blocking: probably useful to explain this matters only if calling `configure_fit()` multiple times.

File: harness/determined/keras/_tf_keras_context.py [hunk 3]
```diff
@@ -40,6 +49,61 @@ def __init__(self, env: det.EnvContext, hvd_config: horovod.HorovodContext):
...
+        self._fit_use_multiprocessing = False
+        self._fit_max_queue_size = 10
+
+    def configure_fit(
+        self,
+        verbose: Optional[bool] = None,
+        class_weight: Any = _arg_not_provided,
+        workers: Optional[int] = None,
+        use_multiprocessing: Optional[bool] = None,
+        max_queue_size: Optional[bool] = None,
```
Comment (aaron276h) [inline harness/determined/keras/_tf_keras_context.py:66]: non-blocking: allow users to specify `validation_steps` also?

File: harness/determined/keras/_tf_keras_context.py [hunk 4]
```diff
@@ -40,6 +49,61 @@ def __init__(self, env: det.EnvContext, hvd_config: horovod.HorovodContext):
...
+              will output periodic text updates rather than the usual ``tqdm`` status bar (which
+              does not render well in trial logs).
+
+           -  ``class_weight``: defaults to ``None``
+
+           -  ``workers``: defaults to 1. Ignored for ``tf.data.dataset`` data.
+
+           -  ``use_multiprocessing``: defaults to `False`. Ignored for ``tf.data.dataset`` data.
+
+           -  ``max_queue_size``: defualts to `10`. Ignored for ``tf.data.dataset`` data.
```
Comment (aaron276h) [inline harness/determined/keras/_tf_keras_context.py:81]: defualts -> defaults

File: harness/determined/keras/_tf_keras_context.py [hunk 5]
```diff
@@ -40,6 +49,61 @@ def __init__(self, env: det.EnvContext, hvd_config: horovod.HorovodContext):
...
+        verbose: Optional[bool] = None,
+        class_weight: Any = _arg_not_provided,
+        workers: Optional[int] = None,
+        use_multiprocessing: Optional[bool] = None,
+        max_queue_size: Optional[bool] = None,
+    ) -> None:
+        """
+        Configure the following arguments to ``model.fit()``:
+
+           -  ``verbose``: defaults to ``True``. The Determined implementation of ``verbose=True``
```
Comment (aaron276h) [inline harness/determined/keras/_tf_keras_context.py:71]: blocking: can we reformat this to be consistent with the other docstrings

File: harness/determined/keras/_tf_keras_context.py [hunk 6]
```diff
@@ -40,6 +49,61 @@ def __init__(self, env: det.EnvContext, hvd_config: horovod.HorovodContext):
...
+           -  ``use_multiprocessing``: defaults to `False`. Ignored for ``tf.data.dataset`` data.
+
+           -  ``max_queue_size``: defualts to `10`. Ignored for ``tf.data.dataset`` data.
+
+        Note that any keyword arguments which are not provided will not overwrite any previously
+        configured settings.
+
+        **Usage Example**
+
+        .. code::
```
Comment (aaron276h) [inline harness/determined/keras/_tf_keras_context.py:88]: `.. code::` -> `.. code-block:: python` so it renders a bit nicer

**Review by aaron276h — APPROVED** [review]
nice work

File: harness/determined/keras/_tf_keras_context.py (hunk 1 above)
Comment (rb-determined-ai) [reply harness/determined/keras/_tf_keras_context.py:94]: I was showing the mechanics of the call.  I could put a comment that explains why it's two different calls?

File: harness/determined/keras/_tf_keras_context.py (hunk 5 above)
Comment (rb-determined-ai) [reply harness/determined/keras/_tf_keras_context.py:71]: you mean swap `-` for `*` or what are you referring to?

File: harness/determined/keras/_tf_keras_context.py (hunk 1 above)
Comment (aaron276h) [reply harness/determined/keras/_tf_keras_context.py:94]: Can you just do this:
```
self.context.configure_fit(verbose=False, workers=5, use_multiprocessing=True)

```

File: harness/determined/keras/_tf_keras_context.py (hunk 5 above)
Comment (aaron276h) [reply harness/determined/keras/_tf_keras_context.py:71]: No reformat it to look like `wrap_dataset()`  (use `Args:`) if you build the docs in the current form, this docstring looks quite different from the other ones

Comment (rb-determined-ai) [reply harness/determined/keras/_tf_keras_context.py:71]: oh, you mean like the `Parameters` format I bet.  Actually, I'm thinking that we should link to the keras docs and call it a day here.  I don't see any reason for documenting anything, except maybe that the output of `verbose` is slightly different.

File: harness/determined/keras/_tf_keras_context.py (hunk 6 above)
Comment (rb-determined-ai) [reply harness/determined/keras/_tf_keras_context.py:88]: done

File: harness/determined/keras/_tf_keras_context.py (hunk 3 above)
Comment (rb-determined-ai) [reply harness/determined/keras/_tf_keras_context.py:66]: let's add it in a future change, after reworking the data loading code.

File: harness/determined/keras/_tf_keras_context.py (hunk 2 above)
Comment (rb-determined-ai) [reply harness/determined/keras/_tf_keras_context.py:83]: done

File: harness/determined/keras/_tf_keras_context.py (hunk 1 above)
Comment (rb-determined-ai) [reply harness/determined/keras/_tf_keras_context.py:94]: why are you opposed to showing that it is safe to call multiple times in the code example?

File: harness/determined/keras/_tf_keras_context.py (hunk 5 above)
Comment (aaron276h) [reply harness/determined/keras/_tf_keras_context.py]: aren't we exposing just a subset of the parameters listed in the keras docs?

File: harness/determined/keras/_tf_keras_context.py (hunk 3 above)
Comment (aaron276h) [reply harness/determined/keras/_tf_keras_context.py:66]: sounds good, do you mind filing a jira so we remember

File: harness/determined/keras/_tf_keras_context.py (hunk 5 above)
Comment (rb-determined-ai) [reply harness/determined/keras/_tf_keras_context.py]: yes, but the subset we expose is clearly listed there without us duplication it in the docstring

File: harness/determined/keras/_tf_keras_context.py (hunk 1 above)
Comment (aaron276h) [reply harness/determined/keras/_tf_keras_context.py:89]: i think just might be a bit confusing, but with your new comment it's fine

File: harness/determined/keras/_tf_keras_context.py (hunk 3 above)
Comment (rb-determined-ai) [reply harness/determined/keras/_tf_keras_context.py:66]: DET-4529
